src/sds.c

```diff
@@ -343,6 +343,48 @@ sds sdsRemoveFreeSpace(sds s) {
     return s;
 }
 
+/* Resize the allocation, this can make the allocation bigger or smaller,
+ * if the size is smaller than currently used len, the data will be truncated */
+sds sdsResize(sds s, size_t size) {
+    void *sh, *newsh;
+    char type, oldtype = s[-1] & SDS_TYPE_MASK;
+    int hdrlen, oldhdrlen = sdsHdrSize(oldtype);
+    size_t len = sdslen(s);
+    sh = (char*)s-oldhdrlen;
+
+    /* Return ASAP if the size is already good. */
+    if (sdsalloc(s) == size) return s;
+
+    /* Truncate len if needed. */
+    if (size < len) len = size;
+
+    /* Check what would be the minimum SDS header that is just good enough to
+     * fit this string. */
+    type = sdsReqType(size);
+    hdrlen = sdsHdrSize(type);
+
+    /* If the type is the same, or at least a large enough type is still
+     * required, we just realloc(), letting the allocator to do the copy
+     * only if really needed. Otherwise if the change is huge, we manually
+     * reallocate the string to use the different header type. */
+    if (oldtype==type || type > SDS_TYPE_8) {
```

Comment:
I don't understand the `type > SDS_TYPE_8` here. The new type might required a different header size (larger or smaller) than the old type. How can we realloc in that case?

Reply:
we avoid changing the type. i.e. we go down from a header with 32bit lengths to a string that doesn't require such a big header, but the excess memory usage doesn't justify the extra work...
notice this branch of the if uses `oldhdrlen` and doesn't modify `s[-1]`

Reply:
This should only happen if `type` is a smaller type than `oldtype`:
```suggestion
    if (oldtype==type || (type > SDS_TYPE_8 && type < oldtype) {
```

Reply:
good catch.. copy paste error..

Reply:
please make a new suggestion with improved comment.